weld/subsystem/src/main/java/org/jboss/as/weld/deployment/processors/ExternalBeanArchiveProcessor.java

```diff
@@ -205,24 +229,121 @@ public void deploy(DeploymentPhaseContext phaseContext) throws DeploymentUnitPro
         }
     }
 
-    private Set<URL> findExportedLocalBeansXml(Module dependencyModule) {
+    /**
+     *
+     * @param beanDiscoveryMode
+     * @param beansXmlUrl
+     * @param indexUrl
+     * @param beanDefiningAnnotations
+     * @return the set of discovered bean classes or null if unable to handle the provided beans.xml url
+     */
+    private Collection<String> discover(BeanDiscoveryMode beanDiscoveryMode, URL beansXmlUrl, URL indexUrl, Set<AnnotationType> beanDefiningAnnotations) {
+        List<String> discoveredClasses = new ArrayList<String>();
+        BiConsumer<String, ClassFile> consumer;
+
+        if (BeanDiscoveryMode.ANNOTATED.equals(beanDiscoveryMode)) {
+            // We must only consider types with bean defining annotations
+            Index index = tryLoadIndex(indexUrl);
+            if (index != null) {
+                // Use the provided index to find ClassInfo
+                consumer = (name, classFile) -> {
+                    ClassInfo classInfo = index.getClassByName(DotName.createSimple(name));
+                    if (classInfo != null && hasBeanDefiningAnnotation(classInfo, beanDefiningAnnotations)) {
+                        discoveredClasses.add(name);
+                    }
+                };
+            } else {
+                // Build ClassInfo on the fly
```

Comment:
We may have already built this and placed it under the org.jboss.as.server.deployment.Attachments#ADDITIONAL_ANNOTATION_INDEXES_BY_MODULE key if annotations are imported, it would be better to re-use this rather than building it again.

Reply:
Hm, but we only build the index if no `META-INF/jandex.idx` is found in exported resources for a specific `beans.xml`. Also it seems `Attachments.ADDITIONAL_ANNOTATION_INDEXES_BY_MODULE` only contains indexes for modules where `META-INF/jandex.idx` exists and annotations are imported.

`Attachments.COMPOSITE_ANNOTATION_INDEX` probably contains indexes from all deployment module dependencies (loaded or calculated on the fly). Shouldn't we use this instead?

Reply:
No, it will build an index if it does not exist (previous versions this was not the case, but we changed this a while ago to remove the need to manually index external modules, although this does only happen when annotations are imported). 

Looking at the current implementation a bit closed I can see this actually won't work though, as it indexes the module as a whole while you are just looking at each jar individually, so the current implementation is fine.

Reply:
> No, it will build...

I think it's built but not added to `Attachments.ADDITIONAL_ANNOTATION_INDEXES_BY_MODULE`, only to `Attachments.COMPOSITE_ANNOTATION_INDEX`:
https://github.com/wildfly/wildfly-core/blob/master/server/src/main/java/org/jboss/as/server/deployment/annotation/CompositeIndexProcessor.java#L101

> ...so the current implementation is fine

OK